packages/devtools_app/lib/src/screens/memory/panes/diff/controller/snapshot_item.dart

```diff
@@ -11,34 +13,56 @@ import '../../../../../shared/memory/heap_graph_loader.dart';
 abstract class SnapshotItem extends DisposableController {
   /// Number to show with auto-generated names that may be non unique, like isolate name.
   int? get displayNumber;
-
-  ValueListenable<bool> get isProcessing => _isProcessing;
-  final _isProcessing = ValueNotifier<bool>(false);
-
-  /// If true, the item contains data, that can be compared and analyzed.
-  bool get hasData;
-
-  @override
-  void dispose() {
-    _isProcessing.dispose();
-    super.dispose();
-  }
 }
 
 class SnapshotDocItem extends SnapshotItem {
   @override
   int? get displayNumber => null;
+}
 
-  @override
-  bool get hasData => false;
+class _Json {
+  static const defaultName = 'defaultName';
+  static const displayNumber = 'displayNumber';
+  static const chunks = 'chunks';
+  static const created = 'created';
+  static const nameOverride = 'nameOverride';
 }
 
 class SnapshotDataItem extends SnapshotItem implements RenamableItem {
   SnapshotDataItem({
-    this.displayNumber,
     required this.defaultName,
-  }) {
-    _isProcessing.value = true;
+    this.displayNumber,
+    this.nameOverride,
+  });
+
+  factory SnapshotDataItem.fromJson(Map<String, dynamic> json) {
+    final result = SnapshotDataItem(
+      displayNumber: json[_Json.displayNumber] as int?,
+      defaultName: json[_Json.defaultName] as String,
+      nameOverride: json[_Json.nameOverride] as String?,
+    );
+
+    final loader = HeapGraphLoaderFromChunks(
+      chunks: json[_Json.chunks] as List<ByteData>,
+      created: json[_Json.created] as DateTime,
+    );
+
+    unawaited(
+      result.loadHeap(loader),
+    ); // Start the loading process, that will result in progress indicator in UI.
+
+    return result;
+  }
+
+  Map<String, dynamic> toJson() {
+    final heap = _heap!; // Not processed heaps are not serializable.
```

Comment:
we do not want to throw exceptions at runtime. Should return an empty map instead?